core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -350,7 +354,8 @@ public TableMetadata updateSchema(Schema newSchema, int newLastColumnId) {
         currentSnapshotId, snapshots, snapshotLog, addPreviousFile(file, lastUpdatedMillis));
   }
 
-  public TableMetadata updatePartitionSpec(PartitionSpec newPartitionSpec) {
+  // newPartitionSpec's partition field IDs should have already been refreshed
+  TableMetadata updatePartitionSpec(PartitionSpec newPartitionSpec) {
```

Comment:
These should still be public like the other methods. Instead of making them private, they should sanity check the spec:
* For a v1 table, field IDs should be assigned from 1000 and should have no gaps
* For a v2 table or newer, all fields should have IDs

Reply:
As this check and field ID change will be in the same PR,  do we still need to check if all fields should have IDs in `newPartitionSpec`?  When the check is called, the PartitionSpec must have a field ID, right?

Reply:
You're right. We just need to check that ids are sequential and starting at 1,000 if the table version is v1.